gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/AndStep.java

```diff
@@ -33,10 +34,18 @@ public AndStep(final Traversal.Admin traversal, final Traversal<S, ?>... travers
 
     @Override
     protected boolean filter(final Traverser.Admin<S> traverser) {
+        GremlinTypeErrorException typeError = null;
         for (final Traversal.Admin<S, ?> traversal : this.traversals) {
-            if (!TraversalUtil.test(traverser, traversal))
-                return false;
+            try {
+                if (!TraversalUtil.test(traverser, traversal))
+                    return false;
+            } catch (GremlinTypeErrorException ex) {
+                // hold onto it until the end in case any other arguments evaluate to FALSE
```

Comment:
I did not understand this. why not short circuit here? (given that the exception is converted to false anyways and the expression will always be false irrespective of the result of other traversals.)

Reply:
Check out the table in the semantics document (in this PR). There is a description of AND/OR/XOR semantics:

`ERROR && FALSE -> FALSE`

`AND` might be wrapped by an operator that handles ERROR differently from FALSE (e.g. `NOT`), so we need to wait on propagating the ERROR until we've seen all the arguments.

Reply:
Understood. Your comment made me think about this change in the semantics itself.

My understanding is that we are changing the semantics to say "illegal comparisons are treated as false i.e. not comparable" unlike before where we used to throw an exception on illegal comparisons. I also understand that the new change in preferred since it removes "null" as a result for a comparison evaluation and beings it closer to equality/equivalence semantics.

But why do we need ternary logic? Why can't Illegal comparison simply be false (instead of error)? (please feel free to point me to a document or discussion thread if this has already been discussed before).

I am advocating for illegal comparison to be treated as `false` at predicate evaluation itself to keep things simpler for customers. It's easy to understand "illegal comparisons are treated as false" but it is difficult to navigate the granular semantics defined in the table in this PR (the one involving XOR).

Reply:
We are trying to bring things more in line with SPARQL, which has well defined logics for error expressions. There are certain value expressions that simply cannot be proved as true or false. There are some nice properties we get by using `ERROR` instead of `FALSE` for these. For example any numeric comparison to NaN is an `ERROR` - P.lt/lte/gt/gte - all `ERROR`, and all reduce to `FALSE` if evaluated on their own. When enclosed by a `NOT`, if we simply immediately used `FALSE` instead of `ERROR`, we would get things like `NOT(P.lt(NaN)) = TRUE`, which is logically inconsistent with `P.gte(NaN) = FALSE`.

It may just be that we need to beef up the documentation.

Reply:
aahh... the `NOT(P.lt(NaN)) = TRUE` is a really good example to explain this. Thank you Mike.